packages/IntegrationTest/tests/InteropCSModuleTests.tsx

```diff
@@ -0,0 +1,180 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT License.
+ *
+ * @format
+ */
+
+import {assert} from 'chai';
+import {NativeModules, NativeEventEmitter} from 'react-native';
+import {functionTest} from './lib/TestDefinition';
+
+const testModuleCs = NativeModules.TestModuleCS;
+
+interface Point {
+  x: number;
+  y: number;
+}
+
+/**
+ * Test for Constants
+ */
+
+functionTest('NumberConstant', () => {
+  const pi = testModuleCs.NumberConstant;
+  assert.isTrue(pi > 3.14 && pi < 3.15);
+});
+
+functionTest('StringConstant', () => {
+  const msg = testModuleCs.StringConstant;
+  assert.equal(msg, 'Hello World');
+});
+
+functionTest('PointConstant', () => {
+  const point = testModuleCs.PointConstant;
+  assert.equal(point.x, 10);
+  assert.equal(point.y, 20);
+  assert.equal(point.z, undefined);
+});
+
+functionTest('ArrayConstant', () => {
+  const arr = testModuleCs.ArrayConstant;
+  assert.equal(arr.length, 3);
+  assert.equal(arr[0], 'one');
+  assert.equal(arr[1], 'two');
+  assert.equal(arr[2], 'three');
+});
+
+functionTest('NullConstant', () => {
+  const n = testModuleCs.NullConstant;
+  assert.equal(n, null);
+});
+
+/**
+ * Tests for Methods
+ */
+
+functionTest('GetAndSetValue', () => {
+  testModuleCs.SetValue(123);
+  testModuleCs.GetValue((result: number) => {
+    assert.equal(result, 123);
+  });
+});
+
+functionTest('GetAndSetValueAsync', async () => {
+  await testModuleCs.SetValueAsync(123);
+  const result = await testModuleCs.GetValueAsync();
+  assert.equal(result, 123);
+});
+
+functionTest('Add', () => {
+  testModuleCs.Add(10, 20, (result: number) => {
+    assert.equal(result, 30);
+  });
+});
+
+functionTest('AddAsync', async () => {
+  const result = await testModuleCs.AddAsync(10, 20);
+  assert.equal(result, 30);
+});
+
+functionTest('Concat', () => {
+  testModuleCs.Concat('abc', 20, (result: string) => {
+    assert.equal(result, 'abc20');
+  });
+});
+
+functionTest('ConcatAsync', async () => {
+  const result = await testModuleCs.ConcatAsync('abc', 20);
+  assert.equal(result, 'abc20');
+});
+
+functionTest('DeltaPoint', () => {
+  testModuleCs.DeltaPoint({x: 1, y: 10}, {x: 10, y: 100}, (result: Point) => {
+    assert.equal(result.x, 9);
+    assert.equal(result.y, 90);
+  });
+});
+
+functionTest('DeltaPointAsync', async () => {
+  const result = await testModuleCs.DeltaPointAsync(
+    {x: 10, y: 100},
+    {x: 1, y: 10},
+  );
+
+  assert.equal(result.x, 9);
+  assert.equal(result.y, 90);
+});
+
+functionTest('Failure', async () => {
+  let seenFailure = false;
+
+  testModuleCs.Failure(
+    () => {
+      assert.fail('This call should have failed');
+    },
+    (e: Error) => {
+      assert.isTrue(e.message.includes('ExpectedFailure'));
+      seenFailure = true;
+    },
+  );
+
+  assert.isTrue(
+    await waitForAsyncResult(() => seenFailure),
```

Comment:
I think we can avoid polling with these tests. Instead of taking an async function, you can return a promise and use `resolve` + `reject` to end the test. This should save a bit of time.

E.g.

```ts
functionTest('Failure', () => new Promise((resolve, reject) => {
  testModuleCs.Failure(
    () => {
      reject(new Error('This call should have failed'));
    },
    (e: Error) => {
      assert.isTrue(e.message.includes('ExpectedFailure'));
      resolve();
    },
  );
}));
```

Reply:
I updated some of the other callback based tests to use Promise too to guarantee order...
We should really advocate for `ReactPromise` in interop as wow this callback api for Native Modules is super ugly...